### ISSUE TITLE
feat: Add Lion optimizer as alternative to AdamW for LoRA training

### ISSUE THREAD

**Issue by VincentGourbin** [VincentGourbin/flux-2-swift-mlx#53]
## Problem

Currently, LoRA training only supports **AdamW** as optimizer (`ResumableAdamW`). While AdamW is a solid default, MLX Swift's `MLXOptimizers` package already ships **Lion** — a newer optimizer discovered by Google Brain (NeurIPS 2023) that offers meaningful advantages for diffusion model fine-tuning.

There is no way for users to choose an alternative optimizer.

## Why Lion?

**Lion** (EvoLved Sign Momentum) was discovered via automated program search ([Symbolic Discovery of Optimization Algorithms, arXiv:2302.06675](https://arxiv.org/abs/2302.06675)). Its update rule is remarkably simple:

```
c = β₁·m + (1 - β₁)·gradient     // interpolation for update direction
m = β₂·m + (1 - β₂)·gradient     // momentum update
param -= lr · sign(c)              // sign-based update (fixed magnitude)
```

### Advantages over AdamW

| Aspect | AdamW | Lion |
|--------|-------|------|
| Optimizer state per param | **2 arrays** (m + v) | **1 array** (m only) |
| Optimizer memory | 2x param size | **1x param size (−50%)** |
| Compute per step | m + v + bias correction | m + sign → **2-15% faster** |
| Diffusion model FID | Baseline | **Better FID** in original paper |
| Generalization gap | 4.2% (CIFAR-10) | **3.8%** |

### Caveats

- **Requires different hyperparameters** — cannot reuse AdamW values
- **Prefers larger batch sizes** (optimal ≥ 64). With batch_size=1, may underperform → recommend using with gradient accumulation
- **sign() gives uniform magnitude updates** — each parameter gets same absolute change regardless of gradient magnitude, which can be suboptimal for LoRA where rank components have different scales
- **Requires careful tuning**, otherwise results are worse than AdamW

### Hyperparameter Translation AdamW → Lion

| Parameter | AdamW (current default) | Lion (recommended) | Ratio |
|-----------|------------------------|-------------------|-------|
| Learning rate | 1e-4 | **1e-5 to 3e-5** | 3-10x smaller |
| Weight decay | 0.01 | **0.03 to 0.1** | 3-10x larger |
| Beta1 | 0.9 | 0.9 | Same |
| Beta2 | 0.999 | **0.99** | Different |
| Epsilon | 1e-8 | N/A | Not needed |

Reference: [Google AutoML Lion README](https://github.com/google/automl/blob/master/lion/README.md)

## Memory Impact

For LoRA training, the optimizer state is a small fraction of total memory (LoRA params ~50-100MB vs transformer activations ~60-90GB), so the absolute memory savings from Lion are modest (~25-50MB). The primary benefits are **speed per step** and potentially **better convergence quality** on diffusion models with sufficient effective batch size.

### 1. Add `OptimizerType` enum

In `LoRATrainingConfig.swift`:

```swift
public enum OptimizerType: String, Codable, CaseIterable, Sendable {
    case adamw
    case lion
}
```

Add `optimizerType` property to `LoRATrainingConfig` (default: `.adamw`).

### 2. Create `ResumableLion` wrapper

In `Sources/Flux2Core/Training/Optimizer/ResumableLion.swift`:

Similar to `ResumableAdamW` but simpler (Lion only has 1 state array per parameter vs 2). Must implement:
- `saveState()` — serialize momentum to checkpoint
- `restoreState()` — restore from checkpoint on resume

### 3. Create optimizer factory

In `SimpleLoRATrainer.swift`, replace the hardcoded `ResumableAdamW` instantiation (line ~374) with a factory:

```swift
let optimizer: any ResumableOptimizer = switch config.optimizerType {
case .adamw:
    ResumableAdamW(learningRate: config.learningRate, betas: (config.adamBeta1, config.adamBeta2),
                   eps: config.adamEpsilon, weightDecay: config.weightDecay)
case .lion:
    ResumableLion(learningRate: config.learningRate, betas: (config.lionBeta1, config.lionBeta2),
                  weightDecay: config.weightDecay)
}
```

This likely requires extracting a `ResumableOptimizer` protocol from `ResumableAdamW`.

### 4. Add YAML configuration

In `TrainingConfigYAML.swift`, add `optimizer` field:

```yaml
training:
  optimizer: lion          # or adamw (default)
  learning_rate: 1e-5      # adjusted for Lion
  weight_decay: 0.1        # adjusted for Lion
```

### 5. Smart defaults per optimizer

When Lion is selected but the user doesn't override LR/WD, apply Lion-appropriate defaults automatically:

| Config | AdamW default | Lion default |
|--------|-------------|-------------|
| learning_rate | 1e-4 | 3e-5 |
| weight_decay | 0.01 | 0.1 |
| beta1 | 0.9 | 0.9 |
| beta2 | 0.999 | 0.99 |

### 6. Files to modify

| File | Change |
|------|--------|
| `LoRATrainingConfig.swift` | Add `OptimizerType` enum, `optimizerType` property, Lion-specific beta defaults |
| `ResumableAdamW.swift` | Extract `ResumableOptimizer` protocol |
| **New** `ResumableLion.swift` | Lion wrapper with checkpoint support |
| `SimpleLoRATrainer.swift` | Optimizer factory based on config |
| `TrainingConfigYAML.swift` | Parse `optimizer` field, Lion-aware defaults |
| `TrainLoRA.swift` (CLI) | Expose `--optimizer` flag |

## Example YAML (Lion)

```yaml
model:
  variant: klein4b

training:
  optimizer: lion
  learning_rate: 3.0e-5
  weight_decay: 0.1
  warmup_steps: 50
  lr_scheduler: cosine
  gradient_accumulation: 4   # recommended: larger effective batch for Lion
  max_steps: 500
```

## Testing Plan

- [ ] Lion converges on Klein 4B with recommended hyperparameters
- [ ] Checkpoint save/restore works for Lion optimizer state
- [ ] AdamW still works unchanged (backward compatibility)
- [ ] Compare loss curves: Lion vs AdamW (same dataset, 250 steps)
- [ ] Validate gradient accumulation improves Lion results (batch_size=1 × accum=4 vs batch_size=1 × accum=1)

## Future Consideration

**Adafactor** could be added as a third option for extreme memory-constrained scenarios (sub-linear memory for weight matrices). Same architecture (enum + protocol + factory) would accommodate it easily.

## References

- [Lion paper: Symbolic Discovery of Optimization Algorithms (arXiv:2302.06675)](https://arxiv.org/abs/2302.06675)
- [Google AutoML Lion README](https://github.com/google/automl/blob/master/lion/README.md)
- [MLX Swift Optimizers source](https://github.com/ml-explore/mlx-swift/blob/main/Source/MLXOptimizers/Optimizers.swift)
- [AdamW vs Lion comparison (SOTAAZ)](https://blog.sotaaz.com/post/adamw-vs-lion-optimizer-comparison-en)
